### PR TITLE
REQUEST FOR COMMENTS: Graceful shutdown of active calls

### PR DESCRIPTION
**Pull request by rmccallum81**
This is an initial step towards a full graceful shutdown of services. Please give comments and ideas for improvements.

### CODE REVIEW

**Comment by icebob** [comment]
Thanks, I will check it...

**Comment by icebob** [comment]
This feature is great, but the extra request storage cuts the performance (from 1.9M rps ro 890k rps). 

**Master benchmark:**
```
===============================
  Moleculer common benchmarks
===============================

Platform info:
==============
   Windows_NT 6.1.7601 x64
   Node.JS: 8.11.0
   V8: 6.2.414.50
   Intel(R) Core(TM) i7-4770K CPU @ 3.50GHz × 8

Suite: Local call
√ broker.call (normal)*             1,975,446 rps
√ broker.call (with params)*        2,136,275 rps

   broker.call (normal)* (#)            0%      (1,975,446 rps)   (avg: 506ns)
   broker.call (with params)*       +8.14%      (2,136,275 rps)   (avg: 468ns)
-----------------------------------------------------------------------

Suite: Call with middlewares
√ No middlewares*        1,987,608 rps
√ 5 middlewares*         1,990,242 rps

   No middlewares* (#)       0%      (1,987,608 rps)   (avg: 503ns)
   5 middlewares*        +0.13%      (1,990,242 rps)   (avg: 502ns)
-----------------------------------------------------------------------

Suite: Call with statistics & metrics
√ No statistics*                    1,934,230 rps
√ With metrics*                       578,381 rps
√ With statistics*                    872,608 rps
√ With metrics & statistics*          418,760 rps

   No statistics* (#)                   0%      (1,934,230 rps)   (avg: 517ns)
   With metrics*                    -70.1%        (578,381 rps)   (avg: 1μs)
   With statistics*                -54.89%        (872,608 rps)   (avg: 1μs)
   With metrics & statistics*      -78.35%        (418,760 rps)   (avg: 2μs)
-----------------------------------------------------------------------

Suite: Remote call with FakeTransporter
√ Remote call echo.reply*           49,791 rps

   Remote call echo.reply*           0%         (49,791 rps)   (avg: 20μs)
-----------------------------------------------------------------------
```

**This PR benchmark:**
```
===============================
  Moleculer common benchmarks
===============================

Platform info:
==============
   Windows_NT 6.1.7601 x64
   Node.JS: 8.11.0
   V8: 6.2.414.50
   Intel(R) Core(TM) i7-4770K CPU @ 3.50GHz × 8

Suite: Local call
√ broker.call (normal)*               890,922 rps
√ broker.call (with params)*          912,201 rps

   broker.call (normal)* (#)            0%        (890,922 rps)   (avg: 1μs)
   broker.call (with params)*       +2.39%        (912,201 rps)   (avg: 1μs)
-----------------------------------------------------------------------

Suite: Call with middlewares
√ No middlewares*          886,702 rps
√ 5 middlewares*           883,858 rps

   No middlewares* (#)       0%        (886,702 rps)   (avg: 1μs)
   5 middlewares*        -0.32%        (883,858 rps)   (avg: 1μs)
-----------------------------------------------------------------------

Suite: Call with statistics & metrics
√ No statistics*                      867,774 rps
√ With metrics*                       389,162 rps
√ With statistics*                    514,174 rps
√ With metrics & statistics*          312,388 rps

   No statistics* (#)                   0%        (867,774 rps)   (avg: 1μs)
   With metrics*                   -55.15%        (389,162 rps)   (avg: 2μs)
   With statistics*                -40.75%        (514,174 rps)   (avg: 1μs)
   With metrics & statistics*         -64%        (312,388 rps)   (avg: 3μs)
-----------------------------------------------------------------------

Suite: Remote call with FakeTransporter
√ Remote call echo.reply*           41,863 rps

   Remote call echo.reply*           0%         (41,863 rps)   (avg: 23μs)
-----------------------------------------------------------------------
```

**Comment by icebob** [comment]
It means if you can't improve the rps above 1.5M, I'll able to merge it only if it's switchable and it's disabled by default.

**Comment by rmccallum81** [comment]
Ok I'm looking at a few options for increased performance. I'll likely make it switchable anyways but the default will be determined by the performance.

**Comment by rmccallum81** [comment]
@icebob I've managed to improve the performance as well as add both a switch (default off) and a performance test

**Comment by zllovesuki** [comment]
Not sure if measuring performance purely based on rps on a local machine is a good metrics. After all, they are microserives not necessary on a single machine. Maybe benchmark it with a networked transporter and see if the performance differs a lot (yes trackContext vs no trackContext)?

**Comment by rmccallum81** [comment]
The benchmark test is to determine the difference in performance based on constructing, storing, recalling, and disposing the context. By doing it locally only you eliminate variances like network latency which would lead to inconsistent results.

**Comment by rmccallum81** [comment]
I looked into the thought @zllovesuki expressed about local vs remote performance with the context tracking. I found that when the FakeTransporter was used there was only a ~6% decrease in performance. Not sure if this is a representation of a need to look into the test regarding the transporter or something else since that benchmark is already significantly slower than the pure local calls. I've added a benchmark test surrounding this scenario.

File: src/service.js [hunk 1]
```diff
@@ -260,6 +279,19 @@ class Service {
...
 	}
 
+	/**
+	 * Retrieve list of active contexts for the service
+	 * @returns {Set}
+	 * @private
+	 * @memberof Service
+	 */
+	_getActiveContexts() {
+		if (_.isFunction(this.broker.ContextFactory.getActiveContexts)) {
```
Comment (icebob) [inline src/service.js]: I don't understand this part... why don't you just call `this.activeContexts` like [here](https://github.com/rmccallum81/moleculer/blob/68d554c4b2cfdb166cb0dbd5e8509488c3366a36/src/context.js#L105)

File: src/service.js [hunk 2]
```diff
@@ -192,17 +198,30 @@ class Service {
...
-				return this.Promise.method(this.schema.stopped).call(this);
-
-			if (Array.isArray(this.schema.stopped)) {
-				return this.schema.stopped
-					.reverse()
-					.map(fn => this.Promise.method(fn.bind(this)))
-					.reduce((p, fn) => p.then(fn), this.Promise.resolve());
-			}
+			return new this.Promise((resolve, reject) => {
+				const timeout = setTimeout(reject, 2000);
```
Comment (icebob) [inline src/service.js]: This 2sec will be a broker option as `gracefulStopTimeout` in milliseconds

File: src/service.js [hunk 3]
```diff
@@ -192,17 +198,30 @@ class Service {
...
+				const checkForContexts = () => {
+					if (this._getActiveContexts().length === 0) {
+						clearTimeout(timeout);
+						resolve();
+					} else {
+						setTimeout(checkForContexts, 100);
+					}
+				};
+				setTimeout(checkForContexts, 0);
```
Comment (icebob) [inline src/service.js]: Please use `setImmediate` instead.

File: src/service.js (hunk 1 above)
Comment (rmccallum81) [reply src/service.js]: Sorry this is a leftover from previous implementation, I'll change it.

**Comment by icebob** [comment]
Why did you close?

**Comment by rmccallum81** [comment]
That was an accident, re-created

**Pull request closed without merging**